### ISSUE TITLE
One region and ivp.detest(testkey)

### ISSUE THREAD

**Issue by componavt** [componavt/math_migration#1]
File: 
newoptimize/optimize_with_4_parameters.py

Line 10:
a = np.array([20.,10.,10.,0.,20.,0. ])#arrive to region,when time is 0,1,2 in 1 and 2 region respectively between 1 b 2 region

1) Я с Вами теперь согласен, что нужно задачу ещё упростить до одного региона. То есть будет два параметра (убывает, привывает) для одного региона.

2) И вторая просьба. Сразу после расчёта средних значений прибывающих и убывающих (строки 19-23), выведите на экран эти значения, например так:

Average number of population, who arrives to region 1 is ТУТ_БУДЕТ_ЧИСЛО
Average number of population, who leaves region 1 is ТУТ_БУДЕТ_ЧИСЛО

3) По поводу функции IVPOPTIM.
В [документации по NodePy 0.6.1](https://media.readthedocs.org/pdf/nodepy/latest/nodepy.pdf) на стр. 20 есть пара невнятных слов про эту функцию.
А вот что это за параметры '2OD', '2OD1' и так далее в Вашем коде?
Не могли бы Вы пояснить мне эти параметры?

Заодно прилагаю рисунок с условиями задачи для удобства работы.

![blackboard_20180306_154744](https://user-images.githubusercontent.com/8916977/41708216-a22258f4-7548-11e8-9cb3-6aa6614f7950.jpg)

**Comment by Inna19** [comment]
2 OD - это название задачи(в данном случае системы из двух дифференциальных уравнений), это название является аргументом для встроенного метода Рунге-Кутты, то есть надо ввести название задачи после вызова метода, чтобы он "знал", что нужно решать. Задача 2OD1 - это система с измененным первым параметром на dq, 2OD2-   система с измененным вторым параметром на dq, соответственно 2OD3 и 2OD4 с измененными 3 и 4 параметрами.  Это все мне нужно для того, чтобы при программировании метода оптимизации найти первые производные по каждому из четырех параметров по формуле (U(a)-U(ai+dq))/dq, где U(a) - решение найденное при заданных начальных параметрах, U(a+dq)-решение,найденное при измененном i-ом параметре, dq-приращение близкое к 0, затем я получаю производную минимизируемой функции по каждом параметру и получаю градиент, затем высчитываю его норму . 2OD11- вторая производная по первому параметру, 2OD12-призводная по первому и второму параметру, все эти производные мне нужны для того, чтобы получить вторую производную минимизируемой функции, а дальше строится матрица Гессе из этих вторых производных. Мне теперь идентифицировать только два параметра?

**Comment by Inna19** [comment]
И с одним уравнением тогда?

**Comment by Inna19** [comment]
Немного ошиблась. 2OD11- задача с измененным первым параметром на 2dq, 2OD12-задача с измененным первым и вторым параметром, все эти задачи мне нужны для того, чтобы получить вторые производные минимизируемой функции, а дальше строится матрица Гессе из этих вторых производных.

**Comment by componavt** [comment]
Подскажите, пожалуйста, где именно есть информация, что такое "2OD12" и подобные названия задач? Можете дать ссылку?

**Comment by Inna19** [comment]
В nodepy есть модуль(или как правильно сказать?) называется ivp.py, в нем содержатся тестовые задачи, которые решаются встроенным численным методом. Поскольку мне нужно было решать свою задачу(которой нет среди тестовых), то я, опираясь на пример программы  ivp.py написала программу для задачи с системой из двух дифференциальных уравнений и назвала её '2 OD' (2  обыкновенных дифференциальных уравнения   в системе) 2OD12 так же по аналогии написана, только там с измененными 1 и 2 параметрами.Пример ivp.py можно увидеть здесь https://github.com/ketch/nodepy/blob/master/nodepy/ivp.py или у себя я открывала python 27 - libs-site-packages-nodepy-ivp.py. Про стандартный набор тестовых задач так же говорится в статье, которую Вы мне сказали разобрать https://sci-hub.la/10.1145/23002.27645, там приводится стандартный набор задач для тестирования численных методов на страницах 19-27. Этот стандартный набор разработчик nodepy использовал и в ivp.py.

**Comment by Inna19** [comment]
Точнее, не программу написала, а написала задачу в программе по аналогии как в ivp.py, а потом численным методом вызывала её, чтобы решить.

**Comment by Inna19** [comment]
Как раз в пособии по nodepy, где говорится про ivp (стр. 20-21), приводится пример решения одной из тестовых задач методом Рунге-Кутты. Эта тестовая задача в ivp.py называется 'nlsin'-система,описывающая колебания линейного осциллятора. И там наверху 20 страницы приводится набор тестовых задач, используемых в ivp.py

**Comment by Inna19** [comment]
Посмотрите,пожалуйста, папку  new_2_parameters. Там программа с одним регионом.

**Comment by componavt** [comment]
1) Запустил в папке math_migration/optimize_with_2_parameters_correction два файла:

- problem_with_new_2_parameters.py
- problem_with_initial_2_parameters.py

Получил рисунки почти идентичные.
![problem_with_initial_2_parameters](https://user-images.githubusercontent.com/8916977/41911438-43c6632a-7955-11e8-81ee-d11e059d09d7.png)
![problem_with_new_2_parameters](https://user-images.githubusercontent.com/8916977/41911439-43ec13fe-7955-11e8-869a-206b7615d6da.png)

Чем эти два файла отличаются? Что такое "новые" параметры? Откуда они получены?

2) В файле problem_with_initial_2_parameters.py вот  строки 63 и 64:
```
u4=np.array([u[1667], u[3334], u[5001]])
t4=np.array([t[1667], t[3334], t[5001]])
```
Откуда взялись эти огромные значения индексов массивов u и массива t? То есть почему именно 1667? Почему не 1668?

А что содержится в предыдущих пяти тысячах ячеек векторов u и t?

**Comment by Inna19** [comment]
Новые параметры - это параметры,полученные методом Левенберга,  в программе optimize_with_2_parameters_correction.py, они выводятся в конце каждой итерации, новые параметры - это те, которые получились в последней итерации. Рисунки с виду мало отличаются, но все же отличаются  вот чем: я  посчитала расстояние между между модельными значениями и фактическими в каждый момент времени при начальных параметрах и при идентифицированных значениях и получилось что расстояние в момент времени 1 на первом рисунке больше, чем на втором, и на первом и на втором это расстояние максимальное. Получается что максимальное расстояние немного  сократилось, но зато немного увеличилось расстояние в момент времени 2 и 3. В программах,содержащих  рисунки эти расстояния выводятся. По поводу u4=np.array([u[1667], u[3334], u[5001]]), t4=np.array([t[1667], t[3334], t[5001]]): дело в том, что в методе Рунге-Кутты я задаю начальный шаг интегрирования, а программа затем сама выбирает нужный шаг, я пыталась разобраться как мне вывести из полученного решения встроенным методом решения в момент времени 1, 2, 3, и определила это опытным путем, то есть t[1667] - это момент времени t= 1,  t[3334]-момент времени t=2, t[5001]]-момент времени t=3, соответственно u[1667] - решение в момент времени 1, u[3334] - решение в момент времени 2, u[5001] - решение в момент времени 3. Если нужно,  могу вывести эти значения в программе, чтобы Вы убедились. Понимаю, что это плохая идея с t4 и u4, у меня по-другому не получилось - буду думать.

**Comment by Inna19** [comment]
В предыдущих пяти тысячах ячеек содержатся решения между моментами времени  0 и 1,1 и 2, 2 и3. Мне для идентификации параметров нужно минимизировать функцию, которая является суммой квадратов разности между модельным и фактическим  решением. То есть мне по идее из полученного встроенным методом решения, нужно выделить то, что получено в моменты времени 1,2,3.

**Comment by componavt** [comment]
Инна, Вы пишите: 
"...пыталась разобраться как мне вывести из полученного решения встроенным методом решения в момент времени 1, 2, 3..."

А Вы не могли бы вывести на график полученные решения не только в трёх точках, но и во всех пяти тысячах точек, которые укладывеются на отрезок времени [1,3]?

И когда построите график для 5000 точек, то постройте график для 50 точек на этом же интервале и оцените: сильно ли изменится точность решения?

"Прилепите", пожалуйста, рисунки-графики к Вашему ответу ниже.

**Comment by Inna19** [comment]
Решения в ранее присланных графиках-это как раз и есть решения во всех 5000 точках(они лежат на  кривых).   А те точки,которые отдельные на рисунках, соединенные пунктиром - это фактические значения. А по поводу попытки вывести точки являющиеся решениями в момент времени 1,2,3  Я верно поняла, что мне на кривых нужно выделить 50 точек?

**Comment by Inna19** [comment]
Точнее, на кривой.

**Comment by componavt** [comment]
>> А по поводу попытки вывести точки являющиеся решениями в момент времени 1,2,3 Я верно поняла, что мне на кривых нужно выделить 50 точек?

Да, верно. Так мы увидим, есть ли смысл делать подробные расчёты на 5000 или всё будет то же самое на 50 точках.
